### PR TITLE
Implement basic fi_write() and fi_read()

### PR DESCRIPTION
**Pull request by ztiffany**
Implement basic fi_write() and fi_read().  Restructure vc->send_queue to be
used for sends ands RMA requests.

Signed-off-by: Zach Tiffany ztiffany@cray.com

@jswaro @sungeunchoi @hppritcha 

This is not ready for commit, I just wanted to get your thoughts on my direction.  Big ideas:
1. Rewrite gnix_ep_push_vc_sendq as _gnix_vc_push_tx_reqs.  This function would now push all TX reqs (sends and RMA).  I renamed vc->send_queue to tx_queue to reflect that the queue contains all types of TX requests.
2. Add _gnix_vc_queue_tx_req() to queue a TX on the VC.  It does the right thing if the network is full or if the VC connection is pending.  It's used on send and RMA requests.
3. Move the code in gnix_ep_send.c into gnix_vc.c.  The two functions that were there are more specific to a VC after I changed gnix_ep_push_vc_sendq.

All criterion tests pass.

Fixes ofi-cray/libfabric-cray#233.
Fixes ofi-cray/libfabric-cray#239.

### CODE REVIEW

**Comment by hppritcha** [comment]
This looks like a good start, I'll not review in super detail till you have it further along.  I do expect the
allocation of the fab req for short messages will impact latency, so we'll need to eventually go back and see how to speed send/recv up.

**Comment by hppritcha** [comment]
bot:retest

**Comment by hppritcha** [comment]
bot:retest

**Comment by hppritcha** [comment]
bot:retest

**Comment by hppritcha** [comment]
bot:retest

**Comment by hppritcha** [comment]
bot:retest

**Comment by hppritcha** [comment]
bot:retest

**Comment by hppritcha** [comment]
using this PR as a test for jenkins changes.  Upgraded ghprb plugin and having problems.

**Comment by hppritcha** [comment]
bot:retest

File: prov/gni/src/gnix_ep.c [hunk 1]
```diff
@@ -474,6 +475,65 @@ static ssize_t gnix_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg,
...
+	gni_return_t status;
+	int ret;
+
+	ep = req->gnix_ep;
+	assert(ep != NULL);
+
+	nic = ep->nic;
+	assert(nic != NULL);
+
+	if (req->len < (16384 - sizeof(struct gnix_smsg_hdr))) {
```
Comment (jswaro) [inline prov/gni/src/gnix_ep.c:493]: Why is 16384 a magic number here?

Comment (hppritcha) [inline prov/gni/src/gnix_ep.c:493]: this will go away eventually, probably as part of the fix for #242

**Comment by ztiffany** [comment]
> I do expect the allocation of the fab req for short messages will impact latency, so we'll need to eventually go back and see how to speed send/recv up.

I'll bet it doesn't make any difference to a micro benchmark after warming up due to the freelist implementation.

File: prov/gni/src/gnix_ep.c (hunk 1 above)
Comment (jswaro) [inline prov/gni/src/gnix_ep.c:493]: Noted, but what is 16384 supposed to be? This should have been a constant defined somewhere that gave some context perhaps. I assume the largest smsg that can be sent is 16384?

**Comment by ztiffany** [comment]
@hppritcha @sungeunchoi @jswaro I've finished and tested a lot of functionality since creating this PR, so it may be useful to get some real review now.  I'll be out for the next week.  If you have time to review before I'm back that would help a lot.  Hopefully the repo doesn't diverge too much before then.

File: prov/gni/src/gnix_ep.c (hunk 1 above)
Comment (hppritcha) [inline prov/gni/src/gnix_ep.c:493]: okay, I'll fix this mystery.

**Comment by ztiffany** [comment]
I added Howard's 5 locking fixes to this branch.  Can we merge this all?

@hppritcha @sungeunchoi @jswaro

**Comment by hppritcha** [comment]
yes please go ahead

**Pull request merged**